Add NodeSets orchestration docs (#1859)
elastic/cloud-on-k8s: docs/orchestration.asciidoc

[id="{p}-orchestration"]

> === NodeSets orchestration

Reviewer: What do we think about calling this just `Orchestration` and mentioning nodesets later? We are talking about nodesets config, but also about stateful sets, behaviors and (as we talked offline) updateStrategy. Mentioning NodeSets feels not broad enough to me.

Author: I can do that. Just "Orchestration" may not be very explicit though. We could also go with `Nodes orchestration`, `Cluster upgrades and orchestration`?

Reviewer 2: 👍 to remove `NodeSets` , I would be in favor of `Nodes Orchestration`

Reviewer 3: +1 for Nodes Orchestration

[id="{p}-nodesets"]
==== NodeSets overview

The Elasticsearch cluster is specified using a list of NodeSets. Each NodeSet represents a group of Elasticsearch nodes sharing the same specification (both Elasticsearch configuration and Kubernetes Pod configuration).

[source,yaml,subs="attributes"]
----
apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: {version}
  nodeSets:

>   - name: master-nodes

Reviewer 4: Nit:
I tend to think this is the name for each node of the set.
I would stay for the name with `master` and `data`, or `master-node` and `data-node` in the singular.
 
```sh
> k get pods
NAME                  READY   STATUS        RESTARTS   AGE
z-es-data-nodes-0     1/1     Running       0          60s
z-es-data-nodes-1     1/1     Running       0          60s
z-es-data-nodes-2     1/1     Running       0          60s
z-es-master-nodes-0   1/1     Running       0          60s
z-es-master-nodes-1   1/1     Running       0          60s
z-es-master-nodes-2   1/1     Running       0          60s
```

Reviewer: I think that you are right that pod name that contains plural is weird looking, but I think I'd still go for it. For me, the `name` reflects the name of the set and the fact we use it it such way for pod names is secondary. I wouldn't dictate the name based on the template we use for pods (eg. if we would have `pod-0-in-master-nodes` pod names it would be fine and singular would look weird).

Reviewer 5: My nit would be more just omitting `-nodes` since it doesn't really add any descriptive value, i.e. just use `- name: masters`. But that is very nitpicky

Author: `masters` looks very nice to me. But `datas` does not?

With only `master` and `data` I'm afraid people will think "this is the way to create a master node". It's not, it's just a name.

Reviewer 5: >masters looks very nice to me. But datas does not?

Data is already plural (datum singular), so I agree. IMO we don't need to overthink this, `master-nodes` or `masters` doesn't make a huge difference for me. We can add a comment that is along the lines of `name: masters # note this is just an arbitrary name and could also be nickelback-is-underrated-fight-me` to maybe help clear up the potential confusion you mentioned

Reviewer 3: `master-node` and `data-node` (singular)
My two cents.

    count: 3
    config:
      master: true
      data: false
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        storageClassName: standard
  - name: data-nodes
    count: 10
    config:
      master: false
      data: true
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1000Gi
        storageClassName: standard
----

> The Elasticcsearch resource above defines two NodeSets: one for master nodes, using 10Gi volumes, and one for data nodes, using 1000Gi volumes. The Elasticsearch cluster will be composed of 13 nodes: 3 master nodes and 10 data nodes.

Reviewer 4: `s|Elasticcsearch|Elasticsearch|`

[id="{p}-upgrading"]
==== Upgrading the cluster

ECK handles smooth upgrades from one cluster specification to another.

Some examples based on the Elasticsearch specification above:

- To add 5 additional data nodes, simply change `count: 10` to `count: 15` in the `data-nodes` NodeSet, then apply the Elasticsearch resource.

> - To increase the RAM memory limit of data Nodes to 32Gi, link:k8s-managing-compute-resources.html[set a different resources limits] in the existing `data-nodes` NodeSet PodTemplate.

Reviewer 4: Small inconsistency: `s|data Nodes|data nodes|` or `Nodes` everywhere.

- To replace dedicated master and dedicated data nodes by nodes having both master and data roles, replace the 2 existing NodeSets by a single one with a different name and the corresponding Elasticsearch configuration settings.

No downtime is expected when NodeSet changes are orchestrated by ECK. It ensures that:

> - data from a node to remove is completely migrated away to other nodes before the node is actually removed

Reviewer 5: ```suggestion
- data on a node that is being removed is completely migrated to other nodes before actually removing said node
```

> - Elasticsearch orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions` correctly account for any change in the cluster

Reviewer 6: are adjusted for any change in the cluster ?

> - Elasticsearch orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions` correctly account for any change in the cluster

Reviewer 5: ```suggestion
- Elasticsearch orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions` are adjusted as necessary for any cluster topology changes
```

> - rolling upgrades are performed in a safe way, reusing PersistentVolumes of the upgraded Elasticsearch nodes

Reviewer 5: ```suggestion
- rolling upgrades are performed safely, reusing PersistentVolumes of the upgraded Elasticsearch nodes (i.e. in-place upgrades rather than grow-and-shrink)
```

I also wonder if it is worth breaking these up into two separate bullets, as we could also do rolling upgrades unsafely while still reusing PVs

Author: "in-place upgrades rather than grow-and-shrink" is our own terminology that we would then need to explain. Not sure that's necessary?

[id="{p}-statefulsets"]
==== StatefulSets orchestration

Behind the scenes, ECK translates each NodeSet specified in the Elasticsearch resource into a link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[StatefulSet in Kubernetes]. The StatefulSet specification is based on the NodeSet specification:

* `count` corresponds to the number of replicas in the StatefulSet, each replica leading to the creation of a Pod, which corresponds to a single Elasticsearch node

> * `podTemplate` is patched with additional required settings to run an Elasticsearch Pod, including the Elasticsearch container and sane default settings for the Pod

Reviewer 6: > sane default settings for the Pod
	
Should we a bit more specific with what we mean by that?

Reviewer 5: ++ to more specificity on defaults, though that may need a link to a separate section so it doesn't take up a lot of space right at the top here. On a more nitpicky note, it is also worth avoiding using "sane" as a modifier, just using "default" or "reasonable default" is okay.

Author: Reworded to something a bit simpler:

>`podTemplate` can be used to specify custom settings for the Elasticsearch Pod, overriding the default ones set by ECK on the generated StatefulSet specification

As you said I think it's not worth going into too many details about what's possible in this section of the doc

* the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name (`clustername-nodesetname`). Each Pod will be assigned the StatefulSet name, suffixed by an ordinal (`clustername-nodesetname-0`). The corresponding Elasticsearch node has the same name.

> The actual Pod creation is handled by the StatefulSet controller in Kubernetes. However, ECK sets the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[StatefulSet update strategy] to `OnDelete`, since it needs full control over when and how Pods get upgraded to a new revision.

Reviewer 5: Agreed on the "However", though we may also be able to remove this whole section as I'm not sure how useful it is for end users.

> The actual Pod creation is handled by the StatefulSet controller in Kubernetes. However, ECK sets the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[StatefulSet update strategy] to `OnDelete`, since it needs full control over when and how Pods get upgraded to a new revision.

Author: I chose to keep it but reword it slightly.

> Whenever a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller ensures PersistentVolumes that were attached to the original Pod are attached again to the replacing Pod.

Reviewer 5: ```suggestion
Whenever a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller ensures PersistentVolumes that were attached to the original Pod are attached again to the newly created Pod.
```

[id="{p}-upgrade-patterns"]
==== Cluster upgrade patterns

> Depending on how the NodeSets specification are updated, ECK handles the Kubernetes resources reconciliation in various ways.

Reviewer 6: Either `NodeSets specifications` or `is updated`

Reviewer 5: Maybe "Depending on how the NodeSets are updated"?

> * When a new NodeSet is added to the Elasticsearch resource, ECK creates the corresponding StatefulSet. It also sets up link:https://kubernetes.io/docs/concepts/configuration/secret/[Secrets] and link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/[ConfigMaps] to hold the Elasticsearch configuration files and TLS certificates.

Reviewer 5: ```suggestion
* When a new NodeSet is added to the Elasticsearch resource, ECK creates the corresponding StatefulSet. It also sets up link:https://kubernetes.io/docs/concepts/configuration/secret/[Secrets] and link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/[ConfigMaps] to hold the TLS certificates and Elasticsearch configuration files.
```

Just to keep them in the same order as they were listed in originally

* When the node count of an existing NodeSet is increased, ECK increases the replicas of the corresponding StatefulSet.
* When the node count of an existing NodeSet is decreased, ECK migrates data away from the corresponding Elasticsearch nodes to remove, then decreases the replicas of the corresponding StatefulSet, once data migration is over. Corresponding <<{p}-volume-claim-templates,PersistentVolumeClaims are automatically removed>>.

> * When an existing NodeSet is removed, ECK migrates data away from the corresponding Elasticsearch nodes to remove, decreases the StatefulSet replicas accordingly, then finally removes the corresponding StatefulSet.

Reviewer 5: It might be worth clarifying if we do it one node at a time or all at once here

Author: Can be both, depending on the updateStrategy.changeBudget. Which has its own section in the doc.

* When the specification of an existing NodeSet is updated (for example the Elasticsearch configuration, or the PodTemplate resources requirements), ECK performs a rolling upgrade of the corresponding Elasticsearch nodes. In order to do so, it follows link:https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html[Elasticsearch rolling upgrade best practices], to slowly upgrade Pods to the newest revision while preventing unavailability of the Elasticsearch cluster. In most cases, it corresponds to restarting Elasticsearch nodes one by one, and reusing the same PersistentVolume data. Note that some <<{p}-orchestration-limitations,cluster topologies may cause the cluster to be unavailable during the upgrade>>.
* When an existing NodeSet is renamed, ECK performs the creation of a new NodeSet with the new name, and the removal of the old NodeSet, according to the NodeSet creation and removal patterns described above. Elasticsearch data is migrated away from the deprecated NodeSet before removal. The Elasticsearch resource <<{p}-update-strategy>> controls how many nodes can exist above or below the target node count during the upgrade.

> In any case, ECK ensures any StatefulSet operation is done in a way that is consistent with the Elasticsearch cluster orchestration best practices, by manipulating orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions`.

Reviewer 5: ```suggestion
In every case, ECK ensures any StatefulSet operation is done in a way that is consistent with the Elasticsearch cluster orchestration best practices, by manipulating orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, and `_cluster/voting_config_exclusions` as necessary.
```

Author: I'm surprised we say "with Elasticsearch cluster". I would have thought it's either "with Elasticsearch" or "with the Elasticsearch cluster"?

Reviewer 6: I think "with Elasticsearch" is fine too, but "with the Elasticsearch cluster" is incorrect in my opinion as the sentence is "consistent with <...> best practices" so it's "Elasticsearch best practices" or "Elasticsearch cluster best practices". I don't know the fancy grammar words for it but I hope it's clear what I mean.

Author: made it "Elasticsearch" only

[id="{p}-orchestration-limitations"]
==== Limitations

Due to the way Kubernetes and StatefulSets behave, ECK orchestration has the following limitations:

> * Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion is not available yet in Kubernetes]. In order to upgrade the storage size, it is however possible to create a new NodeSet, or rename the existing one. The rename will lead to the creation of a new StatefulSet with the specified storage size. The deprecated StatefulSet is removed once Elasticsearch data has been migrated away to nodes of the replacing one, automatically handled by ECK.

Reviewer 6: > The deprecated StatefulSet is removed once Elasticsearch data has been migrated away to nodes of the replacing one, automatically handled by ECK.

Had to read it twice to understand. Do we need the `automatically handled by ECK` bit?

Author: Rephrased to 
> Elasticsearch data from the initial NodeSet will be migrated away to nodes of the replacing one

> * Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion is not available yet in Kubernetes]. In order to upgrade the storage size, it is however possible to create a new NodeSet, or rename the existing one. The rename will lead to the creation of a new StatefulSet with the specified storage size. The deprecated StatefulSet is removed once Elasticsearch data has been migrated away to nodes of the replacing one, automatically handled by ECK.

Reviewer 5: ```suggestion
* Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion is not available yet in Kubernetes]. In order to upgrade the storage size, it is however possible to create a new NodeSet, or rename the existing one. Renaming a NodeSet will lead to the creation of a new StatefulSet with the specified storage size. The original StatefulSet is removed once Elasticsearch data has been migrated away to the nodes of the new StatefulSet.
```

* Cluster availability cannot be guaranteed during the rolling upgrade of a one-node cluster, or for a cluster having indices with no replicas. If an Elasticsearch node holds the only copy of a shard, this shard becomes unavailable while the node is upgraded. Clusters with more than one node and at least one replica per index are considered best practice.
* Elasticsearch Pods may stay `Pending` during a rolling upgrade if the Kubernetes scheduler cannot re-schedule them back. This is especially important when using local PersistentVolumes. If the Kubernetes node bound to a local PersistentVolume does not have enough capacity to host an upgraded Pod which was temporarily removed, that Pod will stay Pending.

> * Rolling upgrades can only progress if the Elasticsearch cluster reports a `green` health. Upgrading a cluster reporting a `yellow` health may cause some shards to become unavailable, turning the cluster health to `red`. To prevent that situation, ECK waits for the cluster to become green in order to make progress. If strictly necessary, it is still possible for the user to manually delete Pods to upgrade. They will be automatically recreated in the newest revision. Note that in some cases where all nodes of a NodeSet are unavailable, ECK may still proceed to upgrading the corresponding Pods, even though the cluster does not report a `green health.`

Reviewer 6: I would get rid of the `a` in front of the colors e.g. "green health" instead of "a green health". 

> If strictly necessary, it is still possible for the user to manually delete Pods to upgrade.

It is still possible for the user to force an upgrade by manually deleting Pods.

> They will be automatically recreated in the newest revision

Not sure if it should be  "They will be automatically recreated at the latest revision" Maybe @anyasabo can help as a native speaker?

Reviewer 7: Not an expert, but "They will be automatically recreated at the latest revision" sounds better. Alternatively, "They will automatically be recreated using the latest revision" could work as well. If I were to write this paragraph, I would write it as:

> Rolling upgrades can only make progress if the Elasticsearch cluster health is `green`. It is risky to attempt upgrading a cluster in the `yellow` state as some shards could become completely unavailable and  degrade the cluster health to `red`. ECK takes the cautionary approach of waiting for `green` before progressing but advanced users may force an upgrade by manually deleting pods themselves. The deleted pods will be automatically recreated at the latest revision. 
> In the exceptional case that all the nodes of a node set are unavailable, the operator might choose to ignore the cluster health and continue with the upgrade anyway.

Reviewer 5: >Note that in some cases where all nodes of a NodeSet are unavailable, ECK may still proceed to upgrading the corresponding Pods, even though the cluster does not report a `green health.`

This can also be changed to "Note that in the event that *all* nodes of a NodeSet..."

Reviewer 5: I should have read everyone else's comments first: agreed with Charith that "at the latest revision" reads better to me. I would also recommend being more specific in the last sentence rather than "may still proceed" or "might choose"

Author: Thanks all!

[id="{p}-orchestration"]
=== NodeSets orchestration

[id="{p}-nodesets"]
==== NodeSets overview

> The Elasticsearch cluster is specified using a list of NodeSets. Each NodeSet represents a group of Elasticsearch nodes sharing the same specification (both Elasticsearch configuration and Kubernetes Pod configuration).

Reviewer 3: Consider using code type font both for `NodeSets` and `NodeSet`.

[source,yaml,subs="attributes"]
----
apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: {version}
  nodeSets:
  - name: master-nodes
    count: 3
    config:
      master: true
      data: false
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        storageClassName: standard
  - name: data-nodes
    count: 10
    config:
      master: false
      data: true
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1000Gi
        storageClassName: standard
----

> The Elasticsearch resource above defines two NodeSets: one for master nodes, using 10Gi volumes, and one for data nodes, using 1000Gi volumes. The Elasticsearch cluster will be composed of 13 nodes: 3 master nodes and 10 data nodes.

Reviewer 3: 1) NodeSets => `NodeSets`
2) For 10Gi and 1000Gi, would it be **storage** instead of **volumes**?
3) Avoid future tense where possible.

Reviewer 5: For 2 it is both, 10Gi is the amount of storage assigned to the volume

[id="{p}-upgrading"]
==== Upgrading the cluster

> ECK handles smooth upgrades from one cluster specification to another.

Reviewer: Should we mention that we can accept new specs at any time and explain how we are always moving towards the current spec? I guess one might think we fully move to previous spec until starting moving to the new one.

Author: Adding `It is possible to apply a new Elasticsearch specification at any time`, which, I think, also solves https://github.com/elastic/cloud-on-k8s/pull/1859#discussion_r331408279.

Some examples based on the Elasticsearch specification above:

> - To add 5 additional data nodes, simply change `count: 10` to `count: 15` in the `data-nodes` NodeSet, then apply the Elasticsearch resource.

Reviewer: That's the only bullet where we talk about applying the resource. Should we skip it or add it above?

Author: I used it as an example for the next ones where it seems "implicit".
I think it's worth mentioning, not sure how to rephrase the above phrase though.

Author: With https://github.com/elastic/cloud-on-k8s/pull/1859#discussion_r331408939 I can safely remove that bit in the first bullet.

- To increase the RAM memory limit of data nodes to 32Gi, link:k8s-managing-compute-resources.html[set a different resources limits] in the existing `data-nodes` NodeSet PodTemplate.
- To replace dedicated master and dedicated data nodes by nodes having both master and data roles, replace the 2 existing NodeSets by a single one with a different name and the corresponding Elasticsearch configuration settings.

> - To upgrade Elasticsearch from version `7.2.0` to `7.3.0`, simply change the value in the `version` field.

Reviewer: Should we mention that downgrades are not supported?

Reviewer 5: I think it would make sense to add to the limitations section, and point out it is an ES limitation.

> - To upgrade Elasticsearch from version `7.2.0` to `7.3.0`, simply change the value in the `version` field.

Reviewer 3: Possible rewording:

> ECK handles smooth upgrades from one cluster specification to another. You can apply a new Elasticsearch specification at any time. 
> 
> Here are a few examples based on the Elasticsearch specification above:
>  
>   - To add five additional data nodes: Change `count: 10` to `count: 15` in the `data-nodes` NodeSet.
>   - To increase the RAM memory limit of data nodes to 32Gi: link:k8s-managing-compute-resources.html[Set a different resources limits] in the existing `data-nodes` NodeSet PodTemplate.
>   - To replace dedicated master and dedicated data nodes by nodes having both master and data roles: Replace the two existing NodeSets by a single one with a different name and the corresponding Elasticsearch configuration settings.
>   - To upgrade Elasticsearch from version `7.2.0` to `7.3.0`: Change the value in the `version` field.

No downtime is expected when NodeSet changes are orchestrated by ECK. It ensures that:

- data on a node that is being removed is completely migrated to other nodes before actually removing said node
- Elasticsearch orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions` are adjusted as necessary for any cluster topology change

> - rolling upgrades are performed safely, reusing PersistentVolumes of the upgraded Elasticsearch nodes

Reviewer 3: Possible rewording:

> ECK orchestrates NodeSet changes with no downtime and makes sure that:
> 
> - Before a node is removed, the relevant data is migrated to other nodes.
> - When a cluster topology changes, the Elasticsearch orchestration settings `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, and `_cluster/voting_config_exclusions` are adjusted accordingly.
>  - Rolling upgrades are performed safely, reusing the `PersistentVolumes` of the upgraded Elasticsearch nodes.

[id="{p}-statefulsets"]
==== StatefulSets orchestration

Behind the scenes, ECK translates each NodeSet specified in the Elasticsearch resource into a link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[StatefulSet in Kubernetes]. The StatefulSet specification is based on the NodeSet specification:

* `count` corresponds to the number of replicas in the StatefulSet, each replica leading to the creation of a Pod, which corresponds to a single Elasticsearch node

> * `podTemplate` can be used to specify custom settings for the Elasticsearch Pod, overriding the default ones set by ECK on the generated StatefulSet specification

Reviewer 2: maybe add a link to https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-pod-template.html

* the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name. Each Pod will be assigned the StatefulSet name suffixed by an ordinal. The corresponding Elasticsearch node has the same name as the Pod.

> The actual Pod creation is handled by the StatefulSet controller in Kubernetes. ECK relies on the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[OnDelete StatefulSet update strategy] since it needs full control over when and how Pods get upgraded to a new revision.

Reviewer 2: Is it worth saying that manually deleting a Pod while an upgrade is in progress is a bad idea and should be done to only unlock a situation where the operator does not seem to make any progress ?

Author: I think we say this at another place:

> ECK takes the cautionary approach of waiting for green health before progressing but advanced users may force an upgrade by manually deleting pods themselves

> Whenever a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller ensures PersistentVolumes that were attached to the original Pod are attached again to the newly created Pod.

Reviewer 3: > When a Pod is removed and recreated (possibly with a newer revision), the StatefulSet controller makes sure that the `PersistentVolumes` attached to the original Pod are then attached to the new Pod.

[id="{p}-upgrade-patterns"]
==== Cluster upgrade patterns

Depending on how the NodeSets are updated, ECK handles the Kubernetes resources reconciliation in various ways.

> * When a new NodeSet is added to the Elasticsearch resource, ECK creates the corresponding StatefulSet. It also sets up link:https://kubernetes.io/docs/concepts/configuration/secret/[Secrets] and link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/[ConfigMaps] to hold the TLS certificates and Elasticsearch configuration files.

Reviewer 2: I think that for advanced users or for support team it would be interesting to say that these actions are done in this exact order.

Author: I think that can also be confusing due to the nature of the reconciliation loop. It may execute step 1 (scale up), then return an error, then skip step 1 because there's nothing else to do and go straight to step 2 (downscale) and return. Then because of the change budget get back to step 1 where we can add more nodes, etc.
It feels hard to explain correctly here?

Reviewer 2: Yes may be that this document is not the right place. I think it is still a valuable information when it's time to understand what the operator is doing and predict what it will do.

* When the specification of an existing NodeSet is updated (for example the Elasticsearch configuration, or the PodTemplate resources requirements), ECK performs a rolling upgrade of the corresponding Elasticsearch nodes. In order to do so, it follows link:https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html[Elasticsearch rolling upgrade best practices], to slowly upgrade Pods to the newest revision while preventing unavailability of the Elasticsearch cluster. In most cases, it corresponds to restarting Elasticsearch nodes one by one, and reusing the same PersistentVolume data. Note that some <<{p}-orchestration-limitations,cluster topologies may cause the cluster to be unavailable during the upgrade>>.
* When an existing NodeSet is renamed, ECK performs the creation of a new NodeSet with the new name, and the removal of the old NodeSet, according to the NodeSet creation and removal patterns described above. Elasticsearch data is migrated away from the deprecated NodeSet before removal. The Elasticsearch resource <<{p}-update-strategy>> controls how many nodes can exist above or below the target node count during the upgrade.

> In every case, ECK ensures any StatefulSet operation is done in a way that is consistent with Elasticsearch orchestration best practices, by manipulating orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, and `_cluster/voting_config_exclusions` as necessary.

Reviewer: s/, and/ and

> In every case, ECK ensures any StatefulSet operation is done in a way that is consistent with Elasticsearch orchestration best practices, by manipulating orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, and `_cluster/voting_config_exclusions` as necessary.

Reviewer 3: In all these cases, ECK handles StatefulSet operations according to the Elasticsearch orchestration best practices, by adjusting the orchestration settings `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, and `_cluster/voting_config_exclusions` accordingly.

[id="{p}-orchestration-limitations"]
==== Limitations

> Due to the way Kubernetes and StatefulSets behave, ECK orchestration has the following limitations:

Reviewer 3: Based on how Kubernetes and StatefulSets operate, ECK orchestration has the following limitations:

> * Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion is not available yet in Kubernetes]. In order to upgrade the storage size, it is however possible to create a new NodeSet, or rename the existing one. Renaming a NodeSet will lead to the creation of a new StatefulSet with the specified storage size. The original StatefulSet is removed once Elasticsearch data has been migrated away to the nodes of the new StatefulSet.

Reviewer 6: I just realised while testing that it is also not possible to add initcontainers to existing NodeSets e.g. to install plugins after the  (also maybe worth  incorporating in the troubleshooting guide @thbkrkr )

Reviewer 3: Possible rewording:

> Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion is not available in Kubernetes yet]. To upgrade the storage size, you can create a new NodeSet, or rename an existing one. Renaming a NodeSet  automatically creates a new StatefulSet with the specified storage size. The original StatefulSet is removed once the Elasticsearch data is migrated to the nodes of the new StatefulSet.

Reviewer 6: >  just realised while testing that it is also not possible to add initcontainers to existing NodeSets 

Poor manual test execution on my part, it is very well possible. Please ignore.

> * Cluster availability cannot be guaranteed during the rolling upgrade of a one-node cluster, or for a cluster having indices with no replicas. If an Elasticsearch node holds the only copy of a shard, this shard becomes unavailable while the node is upgraded. Clusters with more than one node and at least one replica per index are considered best practice.

Reviewer: s/, or/ or

Author: not sure about that one. The comma marks a "pause" in the sentence? (@alaudazzi?)

Reviewer: Ah, I'm not sure either :)

Reviewer 5: Either is fine as far as I know, we can also make it a bulleted list to make it clearer

* Elasticsearch Pods may stay `Pending` during a rolling upgrade if the Kubernetes scheduler cannot re-schedule them back. This is especially important when using local PersistentVolumes. If the Kubernetes node bound to a local PersistentVolume does not have enough capacity to host an upgraded Pod which was temporarily removed, that Pod will stay Pending.

> * Rolling upgrades can only make progress if the Elasticsearch cluster health is green. It is risky to attempt upgrading a cluster in the yellow state as some shards could become completely unavailable and degrade the cluster health to red. ECK takes the cautionary approach of waiting for green before progressing but advanced users may force an upgrade by manually deleting pods themselves. The deleted pods will be automatically recreated at the latest revision. In the exceptional case that all the nodes of a NodeSet are unavailable, probably caused by a misconfiguration, the operator will ignore the cluster health and continue with the upgrade anyway.

Reviewer: s/waiting for green/waiting for green health? or for cluster to become healthy?

> * Rolling upgrades can only make progress if the Elasticsearch cluster health is green. It is risky to attempt upgrading a cluster in the yellow state as some shards could become completely unavailable and degrade the cluster health to red. ECK takes the cautionary approach of waiting for green health before progressing but advanced users may force an upgrade by manually deleting pods themselves. The deleted pods will be automatically recreated at the latest revision. In the exceptional case that all the nodes of a NodeSet are unavailable, probably caused by a misconfiguration, the operator will ignore the cluster health and continue with the upgrade anyway.

Reviewer 2: If the cluster is not green unhealthy Pods (from a K8S pov + not joined ES cluster) can be restarted to unlock some situations.

> * Rolling upgrades can only make progress if the Elasticsearch cluster health is green. It is risky to attempt upgrading a cluster in the yellow state as some shards could become completely unavailable and degrade the cluster health to red. ECK takes the cautionary approach of waiting for green health before progressing but advanced users may force an upgrade by manually deleting pods themselves. The deleted pods will be automatically recreated at the latest revision. In the exceptional case that all the nodes of a NodeSet are unavailable, probably caused by a misconfiguration, the operator will ignore the cluster health and continue with the upgrade anyway.

Reviewer 3: This sentence is a bit confusing:

The deleted pods will be automatically recreated at the latest revision. In the exceptional case that all the nodes of a NodeSet are unavailable, probably caused by a misconfiguration, the operator will ignore the cluster health and continue with the upgrade anyway.

I tried to reword it, but I am not sure if it's correct:

> The deleted pods are automatically recreated at the latest revision. If all the nodes of a NodeSet become unavailable, the operator ignores the cluster health and continues the upgrade.
